File: app/api/v1/model_comparison.py
"""Model Comparison API Endpoints (DEV-256).

Provides REST API endpoints for SUPER_USERs to:
- Run multi-model LLM comparisons
- Vote for best responses (Elo rating)
- Manage model preferences
- View leaderboard and statistics
"""

from datetime import datetime

from fastapi import APIRouter, Depends, HTTPException, Query, status
from sqlalchemy.ext.asyncio import AsyncSession

from app.api.v1.auth import get_current_user
from app.core.logging import logger
from app.models.database import get_db
from app.models.user import User, UserRole
from app.schemas.comparison import (
    AvailableModelsResponse,
    ComparisonRequest,
    ComparisonResponse,
    ComparisonStatsResponse,
    ComparisonWithExistingRequest,
    CreatePendingComparisonRequest,
    LeaderboardResponse,
    ModelPreferencesRequest,
    PendingComparisonData,
    PendingComparisonResponse,
    VoteRequest,
    VoteResponse,
)
from app.services.comparison_service import get_comparison_service

router = APIRouter(prefix="/model-comparison", tags=["Model Comparison"])


def _require_super_user(user: User) -> None:
    """Validate user has SUPER_USER or ADMIN role.

    Args:
        user: Authenticated user

    Raises:
        HTTPException: 403 if user lacks required role
    """
    if user.role not in [UserRole.SUPER_USER.value, UserRole.ADMIN.value]:
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail="Accesso non autorizzato",
        )


@router.post("/compare", response_model=ComparisonResponse)
async def run_comparison(
    request: ComparisonRequest,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
) -> ComparisonResponse:
    """Run a multi-model comparison.

    Sends the same query to multiple LLM models in parallel and returns
    all responses for comparison.

    Args:
        request: Comparison request with query and optional model_ids
        current_user: Authenticated user
        db: Database session

    Returns:
        ComparisonResponse with all model responses

    Raises:
        HTTPException: 403 if not SUPER_USER, 400/500 on validation/execution errors
    """
    _require_super_user(current_user)

    service = get_comparison_service()

    try:
        result = await service.run_comparison(
            query=request.query,
            user_id=current_user.id,
            db=db,
            model_ids=request.model_ids,
        )
        return result
    except ValueError as e:
        error_msg = str(e)
        logger.warning(
            "comparison_validation_error",
            user_id=current_user.id,
            error=error_msg,
        )
        # Determine appropriate status code based on error message
        if "fallito" in error_msg.lower():
            raise HTTPException(
                status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
                detail=error_msg,
            )
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail=error_msg,
        )


@router.post("/compare-with-existing", response_model=ComparisonResponse)
async def run_comparison_with_existing(
    request: ComparisonWithExistingRequest,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
) -> ComparisonResponse:
    """Run comparison using an existing response for the current model.

    This endpoint avoids re-calling the current model by reusing the response
    already obtained in the main chat. Only calls the 4 other best models,
    saving cost on the current model.

    Args:
        request: Request with query and existing response from main chat
        current_user: Authenticated user
        db: Database session

    Returns:
        ComparisonResponse with all model responses (existing + new)

    Raises:
        HTTPException: 403 if not SUPER_USER, 400/500 on validation/execution errors
    """
    _require_super_user(current_user)

    service = get_comparison_service()

    try:
        result = await service.run_comparison_with_existing(
            query=request.query,
            user_id=current_user.id,
            db=db,
            existing_response=request.existing_response,
            enriched_prompt=request.enriched_prompt,
            model_ids=request.model_ids,
        )
        return result
    except ValueError as e:
        error_msg = str(e)
        logger.warning(
            "comparison_with_existing_validation_error",
            user_id=current_user.id,
            error=error_msg,
        )
        # Determine appropriate status code based on error message
        if "fallito" in error_msg.lower():
            raise HTTPException(
                status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
                detail=error_msg,
            )
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail=error_msg,
        )


@router.post("/vote", response_model=VoteResponse)
async def submit_vote(
    request: VoteRequest,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
) -> VoteResponse:
    """Submit a vote for the best model in a comparison.

    Updates Elo ratings based on the vote.

    Args:
        request: Vote request with batch_id and winner_model_id
        current_user: Authenticated user
        db: Database session

    Returns:
        VoteResponse with success status and Elo changes

    Raises:
        HTTPException: 403 if not SUPER_USER, 400/404/409 on validation errors
    """
    _require_super_user(current_user)

    service = get_comparison_service()

    try:
        result = await service.submit_vote(
            batch_id=request.batch_id,
            winner_model_id=request.winner_model_id,
            user_id=current_user.id,
            db=db,
            comment=request.comment,
        )
        return result
    except ValueError as e:
        error_msg = str(e)
        logger.warning(
            "vote_validation_error",
            user_id=current_user.id,
            batch_id=request.batch_id,
            error=error_msg,
        )

        # Determine appropriate status code
        if "non trovata" in error_msg.lower():
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail=error_msg,
            )
        if "già registrato" in error_msg.lower():
            raise HTTPException(
                status_code=status.HTTP_409_CONFLICT,
                detail=error_msg,
            )
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail=error_msg,
        )


@router.get("/models", response_model=AvailableModelsResponse)
async def get_models(
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
) -> AvailableModelsResponse:
    """Get all available models with user preferences.

    Returns the list of all configured models with their Elo ratings
    and whether they are enabled for the current user.

    Args:
        current_user: Authenticated user
        db: Database session

    Returns:
        AvailableModelsResponse with list of models
    """
    _require_super_user(current_user)

    service = get_comparison_service()
    models = await service.get_available_models(current_user.id, db)

    return AvailableModelsResponse(models=models)


@router.put("/models/preferences")
async def update_preferences(
    request: ModelPreferencesRequest,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
) -> dict:
    """Update user model preferences.

    Sets which models are enabled for the user's comparisons.

    Args:
        request: Preferences request with enabled_model_ids
        current_user: Authenticated user
        db: Database session

    Returns:
        Success message
    """
    _require_super_user(current_user)

    service = get_comparison_service()
    await service.update_preferences(
        user_id=current_user.id,
        enabled_model_ids=request.enabled_model_ids,
        db=db,
    )

    logger.info(
        "preferences_updated",
        user_id=current_user.id,
        enabled_count=len(request.enabled_model_ids),
    )

    return {"message": "Preferenze aggiornate con successo"}


@router.get("/stats", response_model=ComparisonStatsResponse)
async def get_stats(
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
) -> ComparisonStatsResponse:
    """Get user comparison statistics.

    Returns stats about the user's comparison activity.

    Args:
        current_user: Authenticated user
        db: Database session

    Returns:
        ComparisonStatsResponse with user statistics
    """
    _require_super_user(current_user)

    service = get_comparison_service()
    stats = await service.get_user_stats(current_user.id, db)

    return ComparisonStatsResponse(stats=stats)


@router.get("/leaderboard", response_model=LeaderboardResponse)
async def get_leaderboard(
    limit: int = Query(default=20, le=100, ge=1),
    db: AsyncSession = Depends(get_db),
) -> LeaderboardResponse:
    """Get the model leaderboard.

    Returns models ranked by Elo rating. This endpoint is public
    (no authentication required) to allow viewing the leaderboard.

    Args:
        limit: Maximum number of results (default 20, max 100)
        db: Database session

    Returns:
        LeaderboardResponse with ranked models
    """
    service = get_comparison_service()
    rankings = await service.get_leaderboard(db, limit=limit)

    return LeaderboardResponse(
        rankings=rankings,
        last_updated=datetime.utcnow(),
    )


@router.post("/pending", response_model=PendingComparisonResponse)
async def create_pending_comparison(
    request: CreatePendingComparisonRequest,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
) -> PendingComparisonResponse:
    """Store pending comparison data from main chat.

    Creates a temporary record with the query and response from main chat,
    allowing the comparison page to retrieve it. Records expire after 1 hour.

    Args:
        request: Request with query, response, model_id, and metrics
        current_user: Authenticated user
        db: Database session

    Returns:
        PendingComparisonResponse with the pending_id UUID

    Raises:
        HTTPException: 403 if not SUPER_USER
    """
    _require_super_user(current_user)

    service = get_comparison_service()
    pending_id = await service.create_pending_comparison(
        user_id=current_user.id,
        query=request.query,
        response=request.response,
        model_id=request.model_id,
        db=db,
        enriched_prompt=request.enriched_prompt,
        latency_ms=request.latency_ms,
        cost_eur=request.cost_eur,
        input_tokens=request.input_tokens,
        output_tokens=request.output_tokens,
        trace_id=request.trace_id,
    )

    logger.info(
        "pending_comparison_created",
        user_id=current_user.id,
        pending_id=pending_id,
        model_id=request.model_id,
        has_enriched_prompt=request.enriched_prompt is not None,
        has_metrics=request.latency_ms is not None,
    )

    return PendingComparisonResponse(pending_id=pending_id)


@router.get("/pending/{pending_id}", response_model=PendingComparisonData)
async def get_pending_comparison(
    pending_id: str,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
) -> PendingComparisonData:
    """Retrieve and delete pending comparison data.

    Retrieves the pending comparison data and deletes it (one-time use).
    Only the user who created the pending comparison can retrieve it.

    Args:
        pending_id: UUID of the pending comparison
        current_user: Authenticated user
        db: Database session

    Returns:
        PendingComparisonData with query, response, and model_id

    Raises:
        HTTPException: 403 if not SUPER_USER, 404 if not found or wrong user
    """
    _require_super_user(current_user)

    service = get_comparison_service()
    data = await service.get_pending_comparison(
        pending_id=pending_id,
        user_id=current_user.id,
        db=db,
    )

    if data is None:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Confronto pendente non trovato",
        )

    logger.info(
        "pending_comparison_retrieved",
        user_id=current_user.id,
        pending_id=pending_id,
    )

    return data

File: src/lib/api/modelComparison.ts
/**
 * Model Comparison API Client (DEV-256)
 *
 * API client for the multi-model LLM comparison feature that allows SUPER_USER
 * role users to compare responses from multiple LLM models.
 */

import {
  ComparisonRequest,
  ComparisonResponse,
  ComparisonWithExistingRequest,
  VoteRequest,
  VoteResponse,
  AvailableModelsResponse,
  LeaderboardResponse,
  ComparisonStatsResponse,
  ModelPreferencesRequest,
  CreatePendingComparisonRequest,
  PendingComparisonResponse,
  PendingComparisonData,
} from '@/types/modelComparison';

/**
 * Base API URL from environment or default
 */
const getBaseUrl = (): string => {
  return process.env.NEXT_PUBLIC_API_URL || 'http://localhost:8000';
};

/**
 * Get authorization token from localStorage
 */
const getAuthToken = (): string | null => {
  if (typeof window === 'undefined') {
    return null;
  }
  return (
    localStorage.getItem('current_session_token') ||
    localStorage.getItem('access_token')
  );
};

/**
 * Make authenticated API request
 */
async function makeRequest<T>(
  endpoint: string,
  options: RequestInit = {}
): Promise<T> {
  const token = getAuthToken();
  if (!token) {
    throw new Error('Non autenticato. Effettua il login per continuare.');
  }

  const url = `${getBaseUrl()}${endpoint}`;
  const headers = new Headers(options.headers);
  headers.set('Authorization', `Bearer ${token}`);
  headers.set('Content-Type', 'application/json');

  try {
    const response = await fetch(url, {
      ...options,
      headers,
    });

    if (!response.ok) {
      const errorData = await response.json().catch(() => ({
        detail: `HTTP ${response.status}: ${response.statusText}`,
      }));
      throw new Error(errorData.detail || 'Errore durante la richiesta API');
    }

    if (response.status === 204) {
      return {} as T;
    }

    return response.json();
  } catch (error) {
    if (
      error instanceof TypeError &&
      error.message.includes('Failed to fetch')
    ) {
      throw new Error(
        'Impossibile connettersi al server. Verifica la tua connessione.'
      );
    }
    throw error;
  }
}

/**
 * Make public API request (no auth required)
 */
async function makePublicRequest<T>(endpoint: string): Promise<T> {
  const url = `${getBaseUrl()}${endpoint}`;

  try {
    const response = await fetch(url, {
      method: 'GET',
      headers: {
        'Content-Type': 'application/json',
      },
    });

    if (!response.ok) {
      const errorData = await response.json().catch(() => ({
        detail: `HTTP ${response.status}: ${response.statusText}`,
      }));
      throw new Error(errorData.detail || 'Errore durante la richiesta API');
    }

    return response.json();
  } catch (error) {
    if (
      error instanceof TypeError &&
      error.message.includes('Failed to fetch')
    ) {
      throw new Error(
        'Impossibile connettersi al server. Verifica la tua connessione.'
      );
    }
    throw error;
  }
}

/**
 * Run a multi-model comparison
 *
 * Sends the same query to multiple LLM models in parallel and returns all responses.
 * Requires SUPER_USER or ADMIN role.
 *
 * @param data - Comparison request with query and optional model_ids
 * @returns ComparisonResponse with all model responses
 */
export async function runComparison(
  data: ComparisonRequest
): Promise<ComparisonResponse> {
  console.log('📤 [Model Comparison] Running comparison:', data.query);

  if (!data.query?.trim()) {
    throw new Error('La domanda non può essere vuota');
  }

  if (data.query.length > 2000) {
    throw new Error('La domanda supera il limite di 2000 caratteri');
  }

  if (data.model_ids && data.model_ids.length < 2) {
    throw new Error('Seleziona almeno 2 modelli per il confronto');
  }

  if (data.model_ids && data.model_ids.length > 6) {
    throw new Error('Massimo 6 modelli per confronto');
  }

  const response = await makeRequest<ComparisonResponse>(
    '/api/v1/model-comparison/compare',
    {
      method: 'POST',
      body: JSON.stringify(data),
    }
  );

  console.log(
    `✅ [Model Comparison] Comparison completed with ${response.responses.length} responses`
  );
  return response;
}

/**
 * Run a comparison with an existing response from main chat
 *
 * Reuses the current model's response from the main chat and only calls
 * the other best models, saving cost on the current model.
 * Requires SUPER_USER or ADMIN role.
 *
 * @param data - Request with query and existing response from main chat
 * @returns ComparisonResponse with all model responses (existing + new)
 */
export async function runComparisonWithExisting(
  data: ComparisonWithExistingRequest
): Promise<ComparisonResponse> {
  console.log(
    '📤 [Model Comparison] Running comparison with existing response:',
    data.query
  );

  if (!data.query?.trim()) {
    throw new Error('La domanda non può essere vuota');
  }

  if (data.query.length > 2000) {
    throw new Error('La domanda supera il limite di 2000 caratteri');
  }

  if (!data.existing_response?.response_text) {
    throw new Error('La risposta esistente non può essere vuota');
  }

  const response = await makeRequest<ComparisonResponse>(
    '/api/v1/model-comparison/compare-with-existing',
    {
      method: 'POST',
      body: JSON.stringify(data),
    }
  );

  console.log(
    `✅ [Model Comparison] Comparison with existing completed with ${response.responses.length} responses`
  );
  return response;
}

/**
 * Submit a vote for the best model in a comparison
 *
 * Updates Elo ratings based on the vote.
 * Requires SUPER_USER or ADMIN role.
 *
 * @param data - Vote request with batch_id and winner_model_id
 * @returns VoteResponse with success status and Elo changes
 */
export async function submitVote(data: VoteRequest): Promise<VoteResponse> {
  console.log('📤 [Model Comparison] Submitting vote:', data);

  if (!data.batch_id) {
    throw new Error('batch_id è obbligatorio');
  }

  if (!data.winner_model_id) {
    throw new Error('Seleziona il modello vincitore');
  }

  const response = await makeRequest<VoteResponse>(
    '/api/v1/model-comparison/vote',
    {
      method: 'POST',
      body: JSON.stringify(data),
    }
  );

  console.log('✅ [Model Comparison] Vote submitted successfully:', response);
  return response;
}

/**
 * Get available models with user preferences
 *
 * Returns the list of all configured models with their Elo ratings
 * and whether they are enabled for the current user.
 * Requires SUPER_USER or ADMIN role.
 *
 * @returns AvailableModelsResponse with list of models
 */
export async function getAvailableModels(): Promise<AvailableModelsResponse> {
  console.log('📥 [Model Comparison] Fetching available models');

  const response = await makeRequest<AvailableModelsResponse>(
    '/api/v1/model-comparison/models',
    {
      method: 'GET',
    }
  );

  console.log(`✅ [Model Comparison] Fetched ${response.models.length} models`);
  return response;
}

/**
 * Update user model preferences
 *
 * Sets which models are enabled for the user's comparisons.
 * Requires SUPER_USER or ADMIN role.
 *
 * @param data - Preferences request with enabled_model_ids
 * @returns Success message
 */
export async function updateModelPreferences(
  data: ModelPreferencesRequest
): Promise<{ message: string }> {
  console.log('📤 [Model Comparison] Updating preferences:', data);

  if (!data.enabled_model_ids || data.enabled_model_ids.length < 2) {
    throw new Error('Seleziona almeno 2 modelli');
  }

  const response = await makeRequest<{ message: string }>(
    '/api/v1/model-comparison/models/preferences',
    {
      method: 'PUT',
      body: JSON.stringify(data),
    }
  );

  console.log('✅ [Model Comparison] Preferences updated:', response);
  return response;
}

/**
 * Get user comparison statistics
 *
 * Returns stats about the user's comparison activity.
 * Requires SUPER_USER or ADMIN role.
 *
 * @returns ComparisonStatsResponse with user statistics
 */
export async function getUserStats(): Promise<ComparisonStatsResponse> {
  console.log('📥 [Model Comparison] Fetching user stats');

  const response = await makeRequest<ComparisonStatsResponse>(
    '/api/v1/model-comparison/stats',
    {
      method: 'GET',
    }
  );

  console.log('✅ [Model Comparison] User stats:', response.stats);
  return response;
}

/**
 * Get the model leaderboard
 *
 * Returns models ranked by Elo rating. This endpoint is public
 * (no authentication required).
 *
 * @param limit - Maximum number of results (default 20, max 100)
 * @returns LeaderboardResponse with ranked models
 */
export async function getLeaderboard(limit = 20): Promise<LeaderboardResponse> {
  console.log('📥 [Model Comparison] Fetching leaderboard');

  const response = await makePublicRequest<LeaderboardResponse>(
    `/api/v1/model-comparison/leaderboard?limit=${Math.min(limit, 100)}`
  );

  console.log(
    `✅ [Model Comparison] Fetched leaderboard with ${response.rankings.length} models`
  );
  return response;
}

/**
 * Create a pending comparison from main chat
 *
 * Stores the query and response from main chat so the comparison page
 * can retrieve it. Records expire after 1 hour.
 * Requires SUPER_USER or ADMIN role.
 *
 * @param data - Request with query, response, and model_id
 * @returns PendingComparisonResponse with pending_id UUID
 */
export async function createPendingComparison(
  data: CreatePendingComparisonRequest
): Promise<PendingComparisonResponse> {
  console.log('📤 [Model Comparison] Creating pending comparison');

  if (!data.query?.trim()) {
    throw new Error('La domanda non può essere vuota');
  }

  if (!data.response?.trim()) {
    throw new Error('La risposta non può essere vuota');
  }

  if (!data.model_id) {
    throw new Error('model_id è obbligatorio');
  }

  const response = await makeRequest<PendingComparisonResponse>(
    '/api/v1/model-comparison/pending',
    {
      method: 'POST',
      body: JSON.stringify(data),
    }
  );

  console.log(
    '✅ [Model Comparison] Pending comparison created:',
    response.pending_id
  );
  return response;
}

/**
 * Get pending comparison data
 *
 * Retrieves the pending comparison data and deletes it (one-time use).
 * Requires SUPER_USER or ADMIN role.
 *
 * @param pendingId - UUID of the pending comparison
 * @returns PendingComparisonData with query, response, and model_id
 */
export async function getPendingComparison(
  pendingId: string
): Promise<PendingComparisonData> {
  console.log('📥 [Model Comparison] Fetching pending comparison:', pendingId);

  const response = await makeRequest<PendingComparisonData>(
    `/api/v1/model-comparison/pending/${pendingId}`,
    {
      method: 'GET',
    }
  );

  console.log('✅ [Model Comparison] Pending comparison retrieved');
  return response;
}
